[JEP-237] Disable usage stats in fips mode

jtnord opened the pull request:
UsageStatistics is so close to being FIPS-140 compliant but fails at the
last hurdle.

We can not encrypt with RSA but we can perform key wrap.  The output does
separate the key and data, by creating a new Key for encrypting the data
and that key is encrypted with RSA.  If only the RSA cipher was created with
Cipher.WRAP_MODE (or Cipher.UNWRAP_MODE) instead of CIPHER.ENCRYPT_MODE (or
CIPHER.DECRYPT_MODE) we would be fine, alas it is not and changing this
would mean a change on the backend and to be able to continue to support
both whilst older versions are in the wild.

The goal of [JEP-237](https://github.com/jenkinsci/jep/blob/master/jep/237/README.adoc) is not to fix all non compliant issues, so here we just ensure that in FIPS mode the usage statistics are disabled.

Downstream of #8482

### screenshots

#### regular

<img width="1125" alt="no-flags" src="https://github.com/jenkinsci/jenkins/assets/494726/f11d96ad-b56d-474b-93b5-a7630d4e2a79">

#### stats disabled by system property

<img width="711" alt="stats_disabled" src="https://github.com/jenkinsci/jenkins/assets/494726/18c7e710-1149-48d0-9f87-c877946a70af">

#### FIPS mode

<img width="880" alt="fips" src="https://github.com/jenkinsci/jenkins/assets/494726/3db64d61-fd20-4091-86d9-0856e318e944">


<!-- Comment:
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue ID you created in Jira.
Note that if you want your changes backported into LTS, you need to create a Jira issue. See https://www.jenkins.io/download/lts/#backporting-process for more information.
-->

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

<!-- Comment:
If the issue is not fully described in Jira, add more information here (justification, pull request links, etc.).

 * We do not require Jira issues for minor improvements.
 * Bug fixes should have a Jira issue to facilitate the backporting process.
 * Major new features should have a Jira issue.
-->

### Testing done

ran the following with and without the FIPS mode flag and validate results are as expected

```
def stats = ExtensionList.lookupSingleton(UsageStatistics.class)
stats.lastAttempt = 0;
println "isDue: " + stats.isDue()
println "isCollected: " + Jenkins.get().isUsageStatisticsCollected()
println "isNoUsageStatistics: " + Jenkins.get().isNoUsageStatistics()
```

round tripped Jenkins config in both modes. (in FIPS we clear (null) the option)

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

### Proposed changelog entries

- Disable anonymous usage statistics when run in FIPS mode.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8483"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

jtnord committed:
[JEP-237] introduce FIPS property

Introduces the code that core and plugins can use to switch behaviour
based on FIPS-140 requirements

jtnord committed:
disable UsageStats in FIPS mode

UsageStatistics is so close to being FIPS-140 compliant but fails at the
last hurdle.

We can not encrypt with RSA but we can perform key wrap.  The output does
separate the key and data, by creating a new Key for encrypting the data
and that key is encrypted with RSA.  If only the RSA cipher was created with
Cipher.WRAP_MODE (or Cipher.UNWRAP_MODE) instead of CIPHER.ENCRYPT_MODE (or
CIPHER.DECRYPT_MODE) we would be fine, alas it is not and changing this
would mean a change on the backend and to be able to continue to support
both whilst older versions are in the wild.

The goal of JEP-237 is not to fix all non compliant issues, so here we
just ensure that in FIPS mode the usage statistics are disabled.

jtnord commented on core/src/main/java/jenkins/security/FIPS140.java:
ignore this file for this review - please see #8482

daniel-beck reviewed (commented):
There are two flags that override the effect of the usage stats option: FIPS (new here) and `DISABLED` (pre-existing). This PR introduces a significant behavior difference between the two (e.g. how it shows up on the UI). This seems unnecessary and doesn't help long-term maintainability. Can this be changed so they're treated similarly?

Note also that JEP-214 telemetry is coupled to the usage statistics submission, so if usage stats are forcibly disabled with `FIPS140#useCompliantAlgorithms` enabled, you'll also no longer get telemetry from those instances.

jtnord committed:
Update core/src/main/java/hudson/model/UsageStatistics.java

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

jtnord commented:
@daniel-beck can you confirm are you asking that the other system property ends up in the same (or similar UX)?

This seems a little at odds with "you'll also no longer get telemetry from those instances" which seemed like a request to decouple the FIPS disabling of stats from the submission of Telemetry, so just want to clarify before making changes.

daniel-beck commented:
> can you confirm are you asking that the other system property ends up in the same (or similar UX)?

I think that would be nice. Right now these flags seem to be presented very differently (e.g. no custom UI indicating it's disabled for `DISABLED`, only FIPS) even though they effectively do the same thing to both usage stats and JEP-214 telemetry. Plus changes like the one to `#isUsageStatisticsCollected` to only account for FIPS, but it is separate from `DISABLED`, are awkward IMO.

> This seems a little at odds with "you'll also no longer get telemetry from those instances" which seemed like a request to decouple the FIPS disabling of stats from the submission of Telemetry, so just want to clarify before making changes.

It's fine for me if that's what the FIPS flag ends up doing, this one is more of a heads up -- no JEP-214 telemetry as implemented, despite it not being needed for compliance with FIPS (AFAICT).

FWIW I can imagine an implementation of the requirement where everything works as before, except https://github.com/jenkinsci/jenkins/blob/c5fbbbe88dc280d0a0876614c0f9aa6fba2ea2af/core/src/main/java/hudson/model/UsageStatistics.java#L104 is changed to never be `true` with FIPS. That should retain JEP-214 telemetry while not submitting usage stats. (Obviously, then the flags work differently, but at least it's deliberately different behavior to retain more functionality even with FIPS enabled, rather than just incidental differences in their presentation).

jtnord commented:
> FWIW I can imagine an implementation of the requirement where everything works as before, except
> 
> https://github.com/jenkinsci/jenkins/blob/c5fbbbe88dc280d0a0876614c0f9aa6fba2ea2af/core/src/main/java/hudson/model/UsageStatistics.java#L104
> 
> is changed to never be `true` with FIPS. That should retain [JEP-214](https://github.com/jenkinsci/jep/tree/master/jep/214) telemetry while not submitting usage stats. (Obviously, then the flags work differently, but at least it's deliberately different behavior to retain more functionality even with FIPS enabled, rather than just incidental differences in their presentation).

ironically that is where I started, but figured I would get some pushback as `FIPS` is more of a "feature" than an individual control flag, so I updated the UI (although I did miss this controlling telemetry at the same time - so thanks for the headsup there)  :)

jtnord committed:
[JEP-237] disable only usage stats but keep telemetry

Update the help and label to accomodate for disabling by system
property, or FIPS.

jtnord commented on core/src/main/java/hudson/model/UsageStatistics.java:
```diff
@@ -102,8 +103,10 @@ public UsageStatistics(String keyImage) {
-        // user opted out. no data collection.
-        if (!Jenkins.get().isUsageStatisticsCollected() || DISABLED)     return false;
+        // user opted out (explicitly or FIPS is requested). no data collection
+        if (!Jenkins.get().isUsageStatisticsCollected() || DISABLED || FIPS140.useCompliantAlgorithms()) {
```
diff from pervious commits is FIPS mode only disabled `UsageStatistics` not Telemetry (which was unaffected)

jtnord commented on core/src/main/java/hudson/model/UsageStatistics.java:
```diff
@@ -212,7 +215,7 @@ public Permission getRequiredGlobalConfigPagePermission() {
         try {
             // for backward compatibility reasons, this configuration is stored in Jenkins
-            Jenkins.get().setNoUsageStatistics(json.has("usageStatisticsCollected") ? null : true);
+            Jenkins.get().setNoUsageStatistics(json.has("usageStatisticsCollected") ? null : Boolean.TRUE);
```
tidy up.  (given this code was updated in previous commits it had been changed) this want's a `Boolean` not a `boolean` so avoids autoboxing and an IDE warning.

jtnord commented on core/src/main/java/jenkins/security/FIPS140.java:
please review https://github.com/jenkinsci/jenkins/pull/8482 not here.

jtnord commented on core/src/main/resources/hudson/model/UsageStatistics/global.properties:
```diff
@@ -22,4 +22,8 @@
 statsBlurb=\
   Help make Jenkins better by sending anonymous usage statistics and crash reports to the Jenkins project
+statsBlurbFIPS=\
+  Help make Jenkins better by sending telemetry to the Jenkins project
```
we only send Telemetry not `UsageStatistics` so the title adapts based on this.

jtnord commented on core/src/main/resources/hudson/model/UsageStatistics/help-usageStatisticsCollected.jelly:
```diff
@@ -6,25 +6,35 @@
-        </p>
-    </div>
+    <j:invokeStatic className="jenkins.security.FIPS140" method="useCompliantAlgorithms" var="fips"/>
+    <j:if test="${!fips}">
```
remove the General usage statistics when in FIPS mode as it is not applicable

jtnord commented on core/src/main/resources/hudson/model/UsageStatistics/help-usageStatisticsCollected.jelly:
```diff
@@ -6,25 +6,35 @@
         <p>
-            In addition to the general usage statistics listed above, the Jenkins project collects telemetry data from specific trials to inform future development.
+            <j:choose>
+                <j:when test="${fips}">
```
adapt to the above section not being present in FIPS mode.

jtnord commented:
> > can you confirm are you asking that the other system property ends up in the same (or similar UX)?
> 
> I think that would be nice. Right now these flags seem to be presented very differently

Feedback should be addressed, see updated screenshots in the description.

jtnord committed:
persist the disabling of stats when disabled by system property

we can no longer toggle the stats option when stats are disabled by
system property.
but this can lead to callers of the API to think that it is enabled.

so if stats are disabled by system property then we disable stats in
the config so users of the API behave appropriately

jtnord commented on core/src/main/java/hudson/model/UsageStatistics.java:
```diff
@@ -212,7 +215,11 @@ public Permission getRequiredGlobalConfigPagePermission() {
         try {
             // for backward compatibility reasons, this configuration is stored in Jenkins
-            Jenkins.get().setNoUsageStatistics(json.has("usageStatisticsCollected") ? null : true);
+            if (DISABLED) {
+                Jenkins.get().setNoUsageStatistics(Boolean.TRUE);
+            } else {
```
persist the disabled nature so that any users of the API get the correct state.

daniel-beck replied:
Unsure about this change.

By default, usage stats submission is enabled (when the field is still `null`), so any callers of the API that ignore the `DISABLED` flag will get the "correct" value only once the global configuration has been saved.

Right now, it's always wrong to rely on the getter alone (depending on use case). This change may tempt callers into ignoring the edge case of "global config has not yet been submitted".

jtnord replied:
> By default, usage stats submission is enabled (when the field is still null), so any callers of the API that ignore the DISABLED flag will get the "correct" value only once the global configuration has been saved.

this is true today, and remains so after this change. 

However if the user previously enabled this usage stats, saved the config and then restarted Jenkins with `-Dhudson.model.UsageStatistics.disabled=true` they (as of this change) have no way to change the value in the UI, so it would forever be persisted as `FALSE`.

in no way does this change that any caller of `Jenkins.isUsageStatsCollected` can get invalid info as the value is `null` until a config change.

daniel-beck replied:
> this is true today, and remains so after this change.

Right, but it's now an edge case more likely to be ignored / callers unaware of.

Don't feel strongly, just seems like something increasing the likelihood of being used wrong (which admittedly was quite substantial even before).

jtnord commented on core/src/main/java/jenkins/security/FIPS140.java:
ignore this file for this review - please see https://github.com/jenkinsci/jenkins/pull/8482

daniel-beck committed:
Merge branch 'master' into disable-usage-stats-in-fips-mode

daniel-beck commented on core/src/main/resources/hudson/model/UsageStatistics/global.properties:
```diff
@@ -22,4 +22,8 @@
+statsBlurbFIPS=\
+  Help make Jenkins better by sending telemetry to the Jenkins project
+disabledBySystemProperty=\
+  The option to send anonymous usage statistics, crash reports and telemetry to the Jenkins project is disabled by a <a href="https://www.jenkins.io/doc/book/managing/system-properties/#hudson-model-usagestatistics-disabled">System Property</a>.
```
```suggestion
  The option to send anonymous usage statistics, crash reports and telemetry to the Jenkins project is disabled by a <a href="https://www.jenkins.io/redirect/usagestatistics-disabled">Java system property</a>.
```

We recommend redirect URLs from Jenkins, as they remain stable in the face of docs reorg (needs corresponding https://github.com/jenkins-infra/jenkins.io PR).

daniel-beck commented on core/src/main/resources/hudson/model/UsageStatistics/global.groovy:
```diff
@@ -1,7 +1,19 @@
+            raw(_("disabledBySystemProperty"))
+        }
+    } else if (FIPS140.useCompliantAlgorithms()) {
+        f.optionalBlock(field: "usageStatisticsCollected", checked: app.usageStatisticsCollected, title: _("statsBlurbFIPS"))
```
It might be useful to be explicit here, that general usage statistics are not sent, e.g. in a `f:description`.

daniel-beck reviewed (approved):
Some nice to have nits left, otherwise looks good. Thanks for making the situation better for `DISABLED`!

timja reviewed (approved):
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!
